target/linux/generic/pending-5.4/780-net-dsa-allow-for-multiple-CPU-ports.patch

```diff
@@ -0,0 +1,212 @@
+CPU ports to user ports in a round robin way, ie. if there are two CPU
+ports connected to eth0 and eth1, and five user ports (lan1..lan5),
+assign them as:
+  lan1 <-> eth0
+  lan2 <-> eth1
+  lan3 <-> eth0
+  lan4 <-> eth1
+  lan5 <-> eth0
+
+Signed-off-by: Marek Behún <marek.behun@nic.cz>
+---
+ include/net/dsa.h |  5 +--
+ net/dsa/dsa2.c    | 82 ++++++++++++++++++++++++++++++-----------------
+ 2 files changed, 56 insertions(+), 31 deletions(-)
+
+--- a/include/net/dsa.h
++++ b/include/net/dsa.h
+@@ -125,9 +125,10 @@ struct dsa_switch_tree {
+ 	struct dsa_platform_data	*pd;
+ 
+ 	/*
+-	 * The switch port to which the CPU is attached.
++	 * The switch ports to which the CPU is attached.
+ 	 */
+-	struct dsa_port		*cpu_dp;
++	size_t			num_cpu_dps;
++	struct dsa_port		*cpu_dps[DSA_MAX_PORTS];
+ 
+ 	/*
+ 	 * Data for the individual switch chips.
+--- a/net/dsa/dsa2.c
++++ b/net/dsa/dsa2.c
+@@ -194,11 +194,12 @@ static bool dsa_tree_setup_routing_table
+ 	return complete;
+ }
+ 
+-static struct dsa_port *dsa_tree_find_first_cpu(struct dsa_switch_tree *dst)
++static void dsa_tree_fill_cpu_ports(struct dsa_switch_tree *dst)
+ {
+ 	struct dsa_switch *ds;
+ 	struct dsa_port *dp;
+ 	int device, port;
++	int count = 0;
+ 
+ 	for (device = 0; device < DSA_MAX_SWITCHES; device++) {
+ 		ds = dst->ds[device];
+@@ -208,28 +209,37 @@ static struct dsa_port *dsa_tree_find_fi
+ 		for (port = 0; port < ds->num_ports; port++) {
+ 			dp = &ds->ports[port];
+ 
+-			if (dsa_port_is_cpu(dp))
+-				return dp;
++			if (dsa_port_is_cpu(dp)) {
++				if (count == ARRAY_SIZE(dst->cpu_dps)) {
++					pr_warn("Tree has too many CPU ports\n");
++					dst->num_cpu_dps = count;
++					return;
++				}
++
++				dst->cpu_dps[count++] = dp;
++			}
+ 		}
+ 	}
+ 
+-	return NULL;
++	dst->num_cpu_dps = count;
+ }
+ 
+-static int dsa_tree_setup_default_cpu(struct dsa_switch_tree *dst)
++static int dsa_tree_setup_default_cpus(struct dsa_switch_tree *dst)
+ {
+ 	struct dsa_switch *ds;
+ 	struct dsa_port *dp;
+-	int device, port;
++	int device, port, i;
+ 
+-	/* DSA currently only supports a single CPU port */
+-	dst->cpu_dp = dsa_tree_find_first_cpu(dst);
+-	if (!dst->cpu_dp) {
++	dsa_tree_fill_cpu_ports(dst);
++	if (!dst->num_cpu_dps) {
+ 		pr_warn("Tree has no master device\n");
+ 		return -EINVAL;
+ 	}
+ 
+-	/* Assign the default CPU port to all ports of the fabric */
++	/* Assign the default CPU port to all ports of the fabric in a round
++	 * robin way. This should work nicely for all sane switch tree designs.
++	 */
++	i = 0;
+ 	for (device = 0; device < DSA_MAX_SWITCHES; device++) {
+ 		ds = dst->ds[device];
+ 		if (!ds)
+@@ -238,18 +248,20 @@ static int dsa_tree_setup_default_cpu(st
+ 		for (port = 0; port < ds->num_ports; port++) {
+ 			dp = &ds->ports[port];
+ 
+-			if (dsa_port_is_user(dp) || dsa_port_is_dsa(dp))
+-				dp->cpu_dp = dst->cpu_dp;
++			if (dsa_port_is_user(dp) || dsa_port_is_dsa(dp)) {
++				dp->cpu_dp = dst->cpu_dps[i++];
```

Comment:
When I test this path, I use a hack here to read the interface's name, and assign CPU ports accordingly.
```diff
-				dp->cpu_dp = dst->cpu_dps[i++];
-				if (i == dst->num_cpu_dps)
-					i = 0;
+				if (dst->num_cpu_dps == 2) {
+					if (strncmp("wan", dp->name, 3) == 0)
+						dp->cpu_dp = dst->cpu_dps[0];
+					else
+						dp->cpu_dp = dst->cpu_dps[1];
+				} else {
+					dp->cpu_dp = dst->cpu_dps[i++];
+					if (i == dst->num_cpu_dps)
+						i = 0;
+				}
```

Specifying the default CPU port in dts may be a better idea.

Reply:
That was rejected upstream. The response was that DTS is only to describe the hardware, not configure it.

Reply:
The preferred way was to configure dsa with ip command.

https://www.spinics.net/lists/netdev/msg595349.html

Reply:
For default configuration, round-robin seemed to us a sane choice. That's what that code does.

Reply:
@frank-w why is trgmii/sgmii better than rgmii? Is RGMII on lower speed?

Reply:
Right, rgmii is 1gbit/s bidirectional. Trgmii is afair 2.3gbit/s and sgmii is 2.5gbit/s. Currently all ports use this gmac and in future lan/wan should be separated (wan rgmii,lan still the faster one).so only wan needs to be changed

Reply:
@frank-w in order to avoid misunderstandings maybe you should specify port speeds next time.
As far as I know, possible speeds on SGMII are 10/100/1000 Mbps. When it is without autonegotiation, it is on 1000 Mbps, it is called 1000base-x. The same protocol, but overclocked by 2.5× is called 2500base-x or HS-SGMII.
So when you wrote SGMII and RGMII, I was thinking about 1000 Mbps on both. This confused me. I have never heard ot TRGMII, though.

Anyway the correct way to do this when CPU is connected to switch via multiple ports of different speeds is
- either to design an algorithm that sets sane defaults according to port speeds (if user ports have all same max speeds and CPU ports have different max speeds, then yes, it would maybe be sane to just take ignore CPU ports with lower speed in the default configuration)
- or to stick with the simple round-robin, and let userspace handle the configuration on those boards which have CPU connected to switch via multiple ports of different speed (who does that? I guess there are not many such boards)

But it is not okay to specify upstream port by hardcoding, nor by DTS.

Reply:
trgmii is similar to sgmii, and yes they are overclocked variants of rgmii.

but only 1 port supports this modes on my boards (on both ends soc+switch)

trgmii on bananapi-r2 (gmac0, switchport 6)
dts: https://elixir.bootlin.com/linux/v5.10-rc6/source/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts#L180
dsa-driver: https://elixir.bootlin.com/linux/v5.10-rc6/source/drivers/net/dsa/mt7530.c#L403

sgmii aka 2500base-x on bananapi-r64 (gmac0, switchport 6)
dts: https://elixir.bootlin.com/linux/v5.10-rc6/source/arch/arm64/boot/dts/mediatek/mt7622-bananapi-bpi-r64.dts#L122
driver:
clock: https://elixir.bootlin.com/linux/v5.10-rc6/source/drivers/net/dsa/mt7530.c#L576
mode itself: https://elixir.bootlin.com/linux/v5.10-rc6/source/drivers/net/dsa/mt7530.c#L2007

i hope i took right positions of sgmii setup in driver-code...i saw also that mt7531 seems to support sgmii on port5 (second upstream-port) too

i mean at initialization time we can choose the first one which is also used at the moment and then let user change it to his needs. I guess all other will be rejected (dts was already and hardcoded is also dirty).

Reply:
@frank-w in the DTS files you sent links for
- on bananpi-r2 there is only one gmac defined in dts (gmac0), so multi-CPU DSA won't affect it.
- on bananapi-r64 there are two gmacs defined (gmac0 and gmac1). gmac0 is used as DSA CPU port on switch port 6. So you are saying that switch port 5 is connected to gmac1, and could also be used as CPU port? If this is so, that IMO you want to use both CPU ports in the default configuration.

The default configuration on bananapi-r64 could look like one of these two (and in either case all user ports are better of than if only one CPU port was used)
```
 gmac0 (2500Mbps)  lan0 lan1 lan2 lan3          wan lan1 lan3
 gmac1 (1000Mbps)  wan                   or     lan0 lan2
```
The first one is basically what you are doing with your patch, but I don't think upstream will like assigning CPU ports by interface labels.

And yes, the round-robin is suboptimal since it can divide the 1000mbps port between 3 user ports (yielding to 333Mbps per port if all were used at once) nad 2500mbps port between 2 user ports. The question is how many such boards are there and whether it makes sense designing a complicated algorithm just because there are 5 boards that need it for their default configuration, when this could be made so that it can be changed from userspace.

Reply:
Yes upstream dts for bpi-r2 has only one gmac defined. I had ported some patches based on openwrt kernel 4.9 from john crispin in my 4.14 and 4.19 tree which adds second gmac.

https://github.com/frank-w/BPI-R2-4.14/blob/4.14-main/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts#L168

Mode is wrong (trgmii not supported on gmac2,but needs to be set there due a driver issue)

There is also an 5.4 approach (phylink) connecting wan-port directly to switch-port 5. But mt7531 does not support the mode directly (only vlan),so with merging this switch (5.10) this is no more possible.

https://github.com/frank-w/BPI-R2-4.14/blob/5.4-gmac/arch/arm/boot/dts/mt7623n-bananapi-bpi-r2.dts#L146

https://github.com/frank-w/BPI-R2-4.14/commit/38f790a805609b255f0cacc097e2df4324a3d041#diff-35d2b0dd263dfe166663a9af8babdaebc5c43f0698cf6c73eb531079e0fbf4da

Reply:
@frank-w well you can try applying these patches and implement `.port_set_cpu_port` method for mt7530 driver, and apply [this](https://lore.kernel.org/netdev/20190824024251.4542-5-marek.behun@nic.cz/) patch to iproute2.

Reply:
I tried applying these patches but first (by patch number 721) failed on 5.10 afair in net/core/dev.c

I guess i need additional patches not upstream

Reply:
I think we can still use device tree to assign _default_ CPU ports, and allow users to change them with iproute2.

Reply:
> I think we can still use device tree to assign _default_ CPU ports, and allow users to change them with iproute2.

You can, but it most probably won't be accepted to upstream kernel.